### PR TITLE
Connect through a proxy with authentication

### PR DESCRIPTION
**Pull request by beugley**
Issue 172 is closed (https://github.com/PyGithub/PyGithub/issues/172), but was never resolved.  Since it's quite old, you may not be interested in this change.  But if you are, this change will allow PyGithub to connect through a proxy that requires authentication.

### CODE REVIEW

File: github/Requester.py [hunk 1]
```diff
@@ -319,7 +320,29 @@ def __createConnection(self):
...
+        ## Connect through a proxy server with authentication, if http_proxy
+        ## set.
+        ## http_proxy: http://user:password@proxy_host:proxy_port
+        ##
+        try:
+           proxy_uri = os.getenv('http_proxy')
+        except:
+           proxy_uri = None
+
+        if (proxy_uri != None):
```
Comment (jzelinskie) [inline github/Requester.py]:
```python
if proxy_uri is not None:
```

File: github/Requester.py [hunk 2]
```diff
@@ -319,7 +320,29 @@ def __createConnection(self):
...
             kwds["timeout"] = self.__timeout  # Did not exist before Python2.6
-        return self.__connectionClass(self.__hostname, self.__port, **kwds)
+
+        ##
+        ## Connect through a proxy server with authentication, if http_proxy
+        ## set.
+        ## http_proxy: http://user:password@proxy_host:proxy_port
+        ##
+        try:
+           proxy_uri = os.getenv('http_proxy')
```
Comment (jzelinskie) [inline github/Requester.py]: is this standard?

File: github/Requester.py (hunk 1 above)
Comment (beugley) [inline github/Requester.py]: No, it's not standard.  I will change it to use "is not".

Comment (jzelinskie) [inline github/Requester.py]: [According to the knowledge card on Google](https://www.google.com/search?q=http_proxy), it is standard. Can we get it in all caps? Like this: `HTTP_PROXY`

File: github/Requester.py (hunk 2 above)
Comment (beugley) [inline github/Requester.py]: http_proxy is respected by many apps as a way to specify a proxy server, I just used the existing convention.

File: github/Requester.py (hunk 1 above)
Comment (beugley) [inline github/Requester.py]: On Linux, it's always in lower case.  How about if I add code to check for both lower and upper case env vars?

Comment (beugley) [inline github/Requester.py]: Please check again.  The code now uses http_proxy, then HTTP_PROXY.  If neither is set, it then connects without proxy.

File: github/Requester.py [hunk 3]
```diff
@@ -319,7 +320,27 @@ def __createConnection(self):
...
         if atLeastPython26:  # pragma no branch (Branch useful only with Python 2.5)
             kwds["timeout"] = self.__timeout  # Did not exist before Python2.6
-        return self.__connectionClass(self.__hostname, self.__port, **kwds)
+
+        ##
+        ## Connect through a proxy server with authentication, if http_proxy
+        ## set.
+        ## http_proxy: http://user:password@proxy_host:proxy_port
+        ##
+        proxy_uri = os.getenv('http_proxy')
```
Comment (jzelinskie) [inline github/Requester.py]:
```python
proxy_uri = os.getenv('http_proxy') or os.getenv('HTTP_PROXY')
```

Comment (beugley) [inline github/Requester.py]: Done.

**Comment by jzelinskie** [comment]
lgtm

**Pull request merged**